Fix saving payment method checkbox availability in classic checkout when signup is disabled

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes STRIPE-465

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR fixes an issue with the Optimized Checkout in the classic checkout when the ability to sign up during checkout is enabled. Previously, the "save payment method" checkbox would always be visible in this case.

The fix consists of adding to the backend and retrieving in the frontend the information about the signup feature being enabled and the user being logged in. Instead of just checking if the saving checkbox exists, we will now check if the user is logged in or logged out but chose to create an account during the checkout.

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout and build this branch on your test environment (`fix/save-card-checkbox-classic-checkout-signup-disabled`)
- Connect your Stripe account to your store
- Enable the OC feature flag (`_wcstripe_feature_oc`). You can do it by either hardcoding the return value of `is_oc_available` to `true` or by running `npm run wp option update _wcstripe_feature_oc 'yes'`
- As a merchant, disable the legacy checkout experience in settings (`wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`)
- Enable the Optimized Checkout

### Logged-in shopper

- As a logged-in shopper, add any product to your cart
- Go to the shortcode checkout page
- Confirm the "save payment method" checkbox is displayed for some payment methods like `cards`

### Logged-out shopper

#### Checkout sign-up enabled

- As a merchant, enable account creation during checkout in WooCommerce > Accounts & Privacy
- As a logged-in shopper, add any product to your cart
- Go to the shortcode checkout page
- Confirm the "save payment method" checkbox is displayed for some payment methods like `cards`, when the "create account" checkbox is selected

#### Checkout sign-up disabled

- As a logged-in shopper, add any product to your cart
- Go to the shortcode checkout page
- Confirm the "save payment method" checkbox is not shown

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Fix saving payment method checkbox availability in classic checkout when signup is disabled

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Fix tests

wjrosa committed:
Merge branch 'develop' into fix/save-card-checkbox-classic-checkout-signup-disabled

wjrosa committed:
Merge branch 'develop' into fix/save-card-checkbox-classic-checkout-signup-disabled

diegocurbelo reviewed (changes requested):
The code looks good, all test cases from the description work as expected:
✔️ Authenticated shopper
✔️ Anonymous shopper + create account during checkout enabled
✔️ Anonymous shopper + create account during checkout disabled

But I found a problem when the shopper already has payment methods saved, the `Save payment method...` is visible when the default payment method is selected (and re-selecting the PM hides it)
![Screenshot 2025-05-23 at 19 07 31](https://github.com/user-attachments/assets/ba5af91b-666d-4e2c-9113-b8393b044e0c)


Also, the title is not shown in the shortcode checkout box, but it's set in the plugin settings:
![Screenshot 2025-05-23 at 19 11 27](https://github.com/user-attachments/assets/1c3fa781-bad5-4fd4-8001-87fcc3ab1455)

wjrosa committed:
Merge branch 'develop' into fix/save-card-checkbox-classic-checkout-signup-disabled

wjrosa committed:
Fix display of saving checkbox when a saved method is selected

wjrosa committed:
Unit test

wjrosa committed:
Fix default OC container title

wjrosa commented:
Thanks for the review, @diegocurbelo ! And for reporting those issues.

> But I found a problem when the shopper already has payment methods saved, the Save payment method... is visible when the default payment method is selected (and re-selecting the PM hides it)

Fixed the issue in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4336/commits/8fa579fc8831dc3e3d5706118aacbac30b8a0132

> Also, the title is not shown in the shortcode checkout box, but it's set in the plugin settings:

Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4336/commits/e65d46e8296c1ef9c18f5bcd69948c5b9f5175b3

diegocurbelo reviewed (changes requested):
Thanks for the fixes @wjrosa!

The checkbox is now correctly hidden by default when the customer has a saved payment method, and shown after selecting "Use new payment method". But it's not working correctly; when clicking any input (like the credit card number), it disappears... and clicking "Use a new payment method" makes it visible again:
https://github.com/user-attachments/assets/0c84a756-288e-4dbe-909a-8ddf175371ae

wjrosa committed:
Merge branch 'develop' into fix/save-card-checkbox-classic-checkout-signup-disabled

wjrosa committed:
Additional check for when the saved payment method is selected

wjrosa commented:
@diegocurbelo I am not able to reproduce the issue in the video, but I added an additional condition to the "saved payment method selected" condition in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4336/commits/541d938505e52da292f86193b5abc52e35c92ead. Can you try again with this update, please?

wjrosa committed:
Merge branch 'develop' into fix/save-card-checkbox-classic-checkout-signup-disabled

diegocurbelo reviewed (approved):
Thanks for the fix @wjrosa!, it works as expected now:

- the checkbox is not present by default
- it shows after selecting Credit Card
- it's hidden when selecting WeChat
- if the checkbox is selected then WeChat is removed from the list

![image (8)](https://github.com/user-attachments/assets/f6ffbc33-6520-4e80-9307-0469622e343e)

wjrosa committed:
Merge branch 'develop' into fix/save-card-checkbox-classic-checkout-signup-disabled
